src/coreclr/src/vm/method.cpp

```diff
@@ -5342,29 +5342,21 @@ LPVOID NDirectMethodDesc::FindEntryPoint(NATIVE_LIBRARY_HANDLE hMod) const
         return reinterpret_cast<LPVOID>(pFunc);
     }
 
-    DWORD probedEntrypointNameLength = (DWORD)(strlen(funcName) + 1); // +1 for charset decorations
-
     // Allocate space for a copy of the entry point name.
+    DWORD probedEntrypointNameLength = (DWORD)(strlen(funcName) + 1); // +1 for charset decorations
     int dstbufsize = (int)(sizeof(char) * (probedEntrypointNameLength + 1)); // +1 for the null terminator
-
     LPSTR szProbedEntrypointName = ((LPSTR)_alloca(dstbufsize));
 
     // Copy the name so we can mangle it.
     strcpy_s(szProbedEntrypointName, dstbufsize, funcName);
-    szProbedEntrypointName[probedEntrypointNameLength] = '\0'; // Add an extra '\0'.
+    szProbedEntrypointName[probedEntrypointNameLength] = '\0'; // Null terminator
+    szProbedEntrypointName[probedEntrypointNameLength - 1] = IsNativeAnsi() ? 'A' : 'W'; // Charset suffix
 
-    if(!IsNativeNoMangled())
+    // Look for entry point with the suffix based on charset
+    FARPROC pProbedFunc = FindEntryPointWithMangling(hMod, szProbedEntrypointName);
```

Comment:
For `IsNativeAnsi`, we will succeed after first probe if we are lucky, and do second probe only some of the time.

For `!IsNativeAnsi`, we will always do two probes and throw away the first one if the second one wins. Can we change so that we will succeed after first probe if we are lucky?

Something like:
```
if (IsNativeNoMangled())
{
    pFunc = FindEntryPointWithMangling(hMod, funcName);
}
else
if (IsIsNativeAnsi())
{
    pFunc = FindEntryPointWithMangling(hMod, funcName);
    if (pFunc == NULL)
       pFunc = FindEntryPointWithSuffix(hMod, funcName, 'A');
}
else
{
    // If it is a unicode function, we are going
    // to need to check for the 'W' API because it takes precedence over the
    // unmangled one (on NT some APIs have unmangled ANSI exports).
    pFunc = FindEntryPointWithSuffix(hMod, funcName, 'W');
    if (pFunc == NULL)
        pFunc = FindEntryPointWithMangling(hMod, funcName);
}
```